text/0035-opentelemetry-protocol.md

```diff
@@ -0,0 +1,425 @@
+
+![Unary](images/otlp-sequential.png)
+
+Sequential operation is recommended when simplicity of implementation is desirable and when the client and the server are connected via very low-latency network, such as for example when the client is an instrumented application and the server is a OpenTelemetry Service running as a local daemon.
+
+The implementations that need to achieve high throughput SHOULD support concurrent Unary calls to achieve higher throughput. The client SHOULD send new requests without waiting for the response to the earlier sent requests, essentially creating a pipeline of requests that are currently in flight that are not acknowledged.
+
+![Streaming](images/otlp-concurrent.png)
+
+The number of concurrent requests SHOULD be configurable.
+
+The maximum achievable throughput is `max_concurrent_requests * max_request_size / (network_latency + server_response_time)`. For example if the request can contain at most 100 spans, network roundtrip latency is 200ms and server response time is 300 ms, then the maximum achievable throughput with one concurrent request is `100 spans / (200ms+300ms)` or 200 spans per second. It is easy to see that in high latency networks or when the server response time is high to achieve good throughput the requests need to be very big or a lot concurrent requests must be done.
+
+If the client is shutting down (e.g. when the containing process wants to exit) the client will optionally wait until all pending acknowledgements are received or until an implementation specific timeout expires. This ensures reliable delivery of telemetry data. The client implementation SHOULD expose an option to turn on and off the waiting during shutdown.
+
+If the client is unable to deliver a certain request (e.g. a timer expired while waiting for acknowledgements) the client SHOULD record the fact that the data was not delivered.
+
+#### Result Code
+
+`Export` response includes a `result_code` field, which indicates whether the server was able to successfully process the received data. Possible values for `result_code` field are:
+
+- `Success` - telemetry data is successfully processed by the server. If the server receives an empty request (a request that does not carry any telemetry data) the server SHOULD respond with `Success`.
+
+- `FailedNotRetryable` - processing of telemetry data failed. The client MUST NOT retry sending the same telemetry data. The telemetry data MUST be dropped. This for example can happen when the request contains bad data and cannot be deserialized or otherwise processed by the server. The client SHOULD maintain a counter of such dropped data.
+
+- `FailedRetryable` - processing of telemetry data failed. The client SHOULD record the error and may retry exporting the same data. This can happen when the server is temporarily unable to process the data.
```

Comment:
it is a good practice to also reply the `retry-after` parameter. For back-off scenarios. Do you think it's worth adding from the beginning or you want to introduce it later?

Reply:
Oh, I should have read further before commenting. Perhaps the only comment here is to mention the next section in this list

Reply:
and perhaps tell that client in case of `FailedRetryable` can retry immediately

Reply:
Correct, when the server wants to ask the client to backoff it returns `Unavailable` error and includes `RetryDelay` data. This approach is borrowed from Google APIs (https://github.com/googleapis/googleapis/blob/c6e62c7e5e61e6dae7fdc3bc3de81f60e6a9445c/google/rpc/error_details.proto#L40) and semantically is very similar to HTTP 503 Service Unavailable with Retry-After response header.

I will add a note that `FailedRetryable` can retry immediately.

Reply:
Added clarification.